### PR TITLE
Fix truncated flags

### PR DESCRIPTION
**Pull request by sandoche2k**
Tese case to test truncated EDNS Responses

### CODE REVIEW

File: docs/specifications/tests/Nameserver-TP/nameserver13.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
+# nameserver13: Test for EDNS truncated response
+
+### test case identifier
```
Comment (matsduf) [inline docs/specifications/tests/Nameserver-TP/nameserver13.md]: Change "##" to "#" here an below.

File: docs/specifications/tests/Nameserver-TP/nameserver13.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+**Nameserver13**
```
Comment (matsduf) [inline docs/specifications/tests/Nameserver-TP/nameserver13.md]: I checked, we use upper case.

`**Nameserver13**` --> `**NAMESERVER13**`

File: docs/specifications/tests/Nameserver-TP/nameserver13.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+### objective
+
+EDNS is a mechanism to announce capabilities of a DNS implementation,
+and is now basically required by any new functionality in DNS such as
+DNSSEC ([rfc 6891]).
+
+[RFC 6891, section 7] states that an OPT record is supposed to be included
+in a truncated response, where-in in the request, an OPT pseudo-RR is present.
```
Comment (matsduf) [inline docs/specifications/tests/Nameserver-TP/nameserver13.md]: Unclear.

File: docs/specifications/tests/Nameserver-TP/nameserver13.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+truncated response with setting buffer size to 512 in the query. We will also
+make sure that the resolution is not retried using TCP.
+
+### Inputs
+"Child Zone" - The domain name to be tested.
+
+### Ordered description of steps to be taken to execute the test case
+
+1. Create a query for the *Child Zone* with DO bit set to '1', a buffer size
+of 512 bytes and with the option for ignoring truncation of UDP responses.
```
Comment (matsduf) [inline docs/specifications/tests/Nameserver-TP/nameserver13.md]: There is no such option unless you would use dig.

File: docs/specifications/tests/Nameserver-TP/nameserver13.md [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+EDNS is a mechanism to announce capabilities of a DNS implementation,
+and is now basically required by any new functionality in DNS such as
+DNSSEC ([rfc 6891]).
+
+[RFC 6891, section 7] states that an OPT record is supposed to be included
+in a truncated response, where-in in the request, an OPT pseudo-RR is present.
+
+In this test case, we will query with DO bit set to '1' and forcing a
+truncated response with setting buffer size to 512 in the query. We will also
+make sure that the resolution is not retried using TCP.
```
Comment (matsduf) [inline docs/specifications/tests/Nameserver-TP/nameserver13.md]: 1. I do not think we should write "we will".
2. How do you make the response being truncated?
3. Retrying with TCP is not done automatically.

File: docs/specifications/tests/Nameserver-TP/nameserver13.md [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+-------
+
+Copyright (c) 2013, 2014, 2015, IIS (The Internet Infrastructure
+Foundation)
Copyright (c) 2013, 2014, 2015, AFNIC
Creative Commons Attribution
+4.0 International License
+
+You should have received a copy of the license along with this work. If not,
+see <https://creativecommons.org/licenses/by/4.0/>.
```
Comment (matsduf) [inline docs/specifications/tests/Nameserver-TP/nameserver13.md]: Remove copyright and license.

File: docs/specifications/tests/Nameserver-TP/nameserver13.md [hunk 7]
```diff
@@ -0,0 +1,87 @@
+# nameserver13: Test for EDNS truncated response
```
Comment (matsduf) [inline docs/specifications/tests/Nameserver-TP/nameserver13.md]: `nameserver13` --> `NAMESERVER13`

File: docs/specifications/tests/Nameserver-TP/nameserver13.md [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+### test case identifier
+
+**Nameserver13**
+
+### objective
```
Comment (matsduf) [inline docs/specifications/tests/Nameserver-TP/nameserver13.md]: "o" --> "O"

File: docs/specifications/tests/Nameserver-TP/nameserver13.md (hunk 3 above)
Comment (sandoche2k) [reply docs/specifications/tests/Nameserver-TP/nameserver13.md]: OPT pseudo-RR is the DO flag.

Comment (matsduf) [reply docs/specifications/tests/Nameserver-TP/nameserver13.md]: Do you mean:

```
[RFC 6891, section 7] states that an OPT record must be included
in a truncated response if the query includes an OPT pseudo record.
```

File: docs/specifications/tests/Nameserver-TP/nameserver13.md (hunk 5 above)
Comment (sandoche2k) [reply docs/specifications/tests/Nameserver-TP/nameserver13.md]: It seems the fallback is automatically to TCP

File: docs/specifications/tests/Nameserver-TP/nameserver13.md (hunk 3 above)
Comment (sandoche2k) [reply docs/specifications/tests/Nameserver-TP/nameserver13.md]: yes

File: docs/specifications/tests/Nameserver-TP/nameserver13.md (hunk 4 above)
Comment (sandoche2k) [reply docs/specifications/tests/Nameserver-TP/nameserver13.md]: I have re-worded the statement. Please check.

File: docs/specifications/tests/Nameserver-TP/nameserver13.md (hunk 5 above)
Comment (sandoche2k) [reply docs/specifications/tests/Nameserver-TP/nameserver13.md]: Text modified. Please check.

File: docs/specifications/tests/Nameserver-TP/nameserver13.md (hunk 6 above)
Comment (sandoche2k) [reply docs/specifications/tests/Nameserver-TP/nameserver13.md]: Done

File: docs/specifications/tests/Nameserver-TP/nameserver13.md [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+None.
+
+[RFC 6891]: https://tools.ietf.org/html/rfc6891
+[RFC 6891, section 7]: https://tools.ietf.org/html/rfc6891#section-7
+[Method4]: ../Methods.md#method-4-delegation-name-server-addresses
+[Method5]: ../Methods.md#method-5-in-zone-addresses-records-of-name-servers
+[NO_RESPONSE]: #outcomes
+[NO_EDNS_SUPPORT]: #outcomes
+[NS_ERROR]: #outcomes
+
```
Comment (matsduf) [inline docs/specifications/tests/Nameserver-TP/nameserver13.md]: NO_TC_FLAG is missing here.

File: docs/specifications/tests/Nameserver-TP/nameserver13.md [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+3. For each name server in *Name Server IP* do:
+	1. Send the query to the name server and collect the response.
+	2. If there is no DNS response, output *[NO_RESPONSE]* and go to
+      next server.
+	3. The DNS response has the RCODE "FORMERR" then output
+      *[NO_EDNS_SUPPORT]* and go to the next server. 
+	4. The DNS response has no 'TC' flag, output *[NO_TC_FLAG]* and go to
+the next server.
+
+4. Else, if the DNS response meet the following criteria,
```
Comment (matsduf) [inline docs/specifications/tests/Nameserver-TP/nameserver13.md]: Indent one step here an below.

File: docs/specifications/tests/Nameserver-TP/nameserver13.md (hunk 9 above)
Comment (sandoche2k) [reply docs/specifications/tests/Nameserver-TP/nameserver13.md]: Added with the option "WARNING". Are you ok with that?

**Comment by sandoche2k** [comment]
@matsduf Modified as per your review. Check again.

File: docs/specifications/tests/Nameserver-TP/nameserver13.md [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+'DO' bit. 
+
+The query will be tested on a DNSSEC signed zone. This test will not give a
+valid result if the zone is not signed.  
+
+In addition, the EDNS buffer size should be set to 512, since most signed DNSKEY
+responses are bigger than 512 bytes. 
+
+The objective  of the query in this test  is to elicit a truncated response
+from the server.
```
Comment (matsduf) [inline docs/specifications/tests/Nameserver-TP/nameserver13.md]: This section is about the objective of the Test Case. The details of the buffer size fits better in the steps, where an explanation is given.

File: docs/specifications/tests/Nameserver-TP/nameserver13.md [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+## Inputs
+
+"Child Zone" - The domain name to be tested.
+
+## Ordered description of steps to be taken to execute the test case
+
+1. Create a query for the *Child Zone* that is signed with 'DO' bit set to '1' and
```
Comment (matsduf) [inline docs/specifications/tests/Nameserver-TP/nameserver13.md]: What should the query be for? ANY?

File: docs/specifications/tests/Nameserver-TP/nameserver13.md [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+   ("Name Server IP").
+
+3. For each name server in *Name Server IP* do:
+	1. Send the query to the name server and collect the response.
+	2. If there is no DNS response, output *[NO_RESPONSE]* and go to
+      	next server.
+	3. Else, if the DNS response has the RCODE "FORMERR" then output
+      	*[NO_EDNS_SUPPORT]* and go to the next server. 
+	4. Else, if the DNS response has no 'TC' flag, output *[NO_TC_FLAG]* and go to
+	the next server.
```
[the remark this answers is not in the document]
Comment (sandoche2k) [reply docs/specifications/tests/Nameserver-TP/nameserver13.md]: It is already a WARNING

File: docs/specifications/tests/Nameserver-TP/nameserver13.md (hunk 11 above)
Comment (sandoche2k) [reply docs/specifications/tests/Nameserver-TP/nameserver13.md]: I feel that the three sentences in the objective gives a better description to the audience

File: docs/specifications/tests/Nameserver-TP/nameserver13.md (hunk 12 above)
Comment (sandoche2k) [reply docs/specifications/tests/Nameserver-TP/nameserver13.md]: DNSKEY

File: docs/specifications/tests/Nameserver-TP/nameserver13.md [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+'DO' bit. 
+
+This Test Case is not applicable unless the zone is DNSSEC signed. No DS records
+are needed in the parent zone
+
+In addition, the EDNS buffer size should be set to 512, since most signed DNSKEY
+responses are bigger than 512 bytes. 
+
+The objective  of the query in this test  is to elicit a truncated response
+from the server.
```
Comment (matsduf) [inline docs/specifications/tests/Nameserver-TP/nameserver13.md]: I suggest the following text under the Objective section:

```
EDNS is a mechanism to announce capabilities of a DNS implementation,
and is now basically required by any new functionality in DNS such as
DNSSEC ([RFC 6891]).

[RFC 6891, section 7] states that an OPT record must be included
in a truncated response if the query includes an OPT pseudo record.

This Test Case will try verify that if the response to a query with an OPT record
is truncated, then the response will contain an OPT record.

To trigger a truncated response the DO bit is set and the buffer size is limited to 512
octets (bytes). If the zone is not signed with DNSSEC the response will probably not
be truncated anyway.
```

File: docs/specifications/tests/Nameserver-TP/nameserver13.md [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+
+## Inputs
+
+"Child Zone" - The domain name to be tested.
+
+## Ordered description of steps to be taken to execute the test case
+
+1. Create a DNSKEY query for the *Child Zone* that is signed with 'DO' bit set to '1' and
+setting the buffer size to 512 bytes
```
[the remark this answers is not in the document]
Comment (sandoche2k) [reply docs/specifications/tests/Nameserver-TP/nameserver13.md]: Minimum is 512 bytes : https://docs.infoblox.com/display/NAG8/Setting+DNSSEC+Parameters

File: docs/specifications/tests/Nameserver-TP/nameserver13.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (sandoche2k) [reply docs/specifications/tests/Nameserver-TP/nameserver13.md]: I also think that this test is unnecessary, but you wanted it :-), in your previous comment.

File: docs/specifications/tests/Nameserver-TP/nameserver13.md [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+	3. Else, if the DNS response has the RCODE "FORMERR" then output
+      	*[NO_EDNS_SUPPORT]* and go to the next server. 
+	4. Else, if the DNS response has no 'TC' flag, output *[NO_TC_FLAG]* and go to
+	the next server.
+	5. Else, if the DNS response meet the following criteria,
+      	then just go to the next name server (no error):
+		1. The DNS response has the RCODE "NOERROR".
+		2. The header contains the 'TC' flag set.
+		3. The pseudo-section response has an OPT record with version set to 0.
+
```
[the remark this answers is not in the document]
Comment (sandoche2k) [reply docs/specifications/tests/Nameserver-TP/nameserver13.md]: Yes, the OPT record with version set to '0' as per the 3.5.3

[the remark this answers is not in the document]
Comment (matsduf) [reply docs/specifications/tests/Nameserver-TP/nameserver13.md]:
```
    5. Else, if the DNS response meet the following criteria output 
        *[MISSING_OPT_IN_TRUNCATED]*:
        1. The DNS response is truncated (the "TC" flag is set).
        2. The DNS response has no OPT record.
```

File: docs/specifications/tests/Nameserver-TP/nameserver13.md [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+with the severity level *WARNING*, but no message with severity level
+*ERROR* or *CRITICAL*.
+
+The outcome of this Test case is "pass" in all other cases.
+
+Message                           | Default severity level (when message is outputed)
+:---------------------------------|:--------------------------------------------------
+NO_RESPONSE                       | WARNING
+NO_EDNS_SUPPORT                   | WARNING
+NS_ERROR			  | WARNING
```
Comment (matsduf) [inline docs/specifications/tests/Nameserver-TP/nameserver13.md]:
```
MISSING_OPT_IN_TRUNCATED   |   WARNING
```

File: docs/specifications/tests/Nameserver-TP/nameserver13.md [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+None.
+
+[RFC 6891]: https://tools.ietf.org/html/rfc6891
+[RFC 6891, section 7]: https://tools.ietf.org/html/rfc6891#section-7
+[Method4]: ../Methods.md#method-4-delegation-name-server-addresses
+[Method5]: ../Methods.md#method-5-in-zone-addresses-records-of-name-servers
+[NO_RESPONSE]: #outcomes
+[NO_EDNS_SUPPORT]: #outcomes
+[NO_TC_FLAG]: #outcomes
+[NS_ERROR]: #outcomes
```
Comment (matsduf) [inline docs/specifications/tests/Nameserver-TP/nameserver13.md]:
```
[MISSING_OPT_IN_TRUNCATED]: #outcomes
```

**Comment by sandoche2k** [comment]
updated

File: docs/specifications/tests/Nameserver-TP/nameserver13.md [hunk 19]
```diff
@@ -0,0 +1,88 @@
+# NAMSERVER13: Test for EDNS truncated response
```
Comment (matsduf) [inline docs/specifications/tests/Nameserver-TP/nameserver13.md]: "EDNS truncated response" sounds strange. 

"truncated response on query with EDNS"
"truncated EDNS response"
"truncated response on EDNS query"

**Comment by sandoche2k** [comment]
Updated

**Pull request merged**